dexios-gui/src/main.rs

```diff
@@ -1,3 +1,451 @@
+                    ui.radio_value(&mut self.encrypt.key, Key::Keyfile, "Keyfile");
+                    if ui
+                        .radio_value(&mut self.encrypt.key, Key::AutoGenerate, "Auto Generate")
+                        .clicked()
+                    {
+                        self.encrypt.autogenerated_passphrase =
+                            gen_passphrase().expose().to_string();
+                    };
+                });
+
+                if self.encrypt.key == Key::Password {
+                    ui.horizontal(|ui| {
+                        ui.label("Password: ");
+                        ui.add(
+                            egui::TextEdit::singleline(&mut self.encrypt.password)
+                                .hint_text("Password: ")
+                                .password(true),
+                        );
+                        ui.add(
+                            egui::TextEdit::singleline(&mut self.encrypt.password_validation)
+                                .hint_text("Password (for validation): ")
+                                .password(true),
+                        );
+                    });
+                }
+
+                if self.encrypt.key == Key::Keyfile {
+                    ui.horizontal(|ui| {
+                        ui.label("Keyfile: ");
+
+                        ui.add(
+                            egui::TextEdit::singleline(&mut self.encrypt.keyfile_path)
+                                .hint_text("Path to the keyfile"),
+                        );
+                        if ui.button("Select File").clicked() {
+                            if let Some(path) = rfd::FileDialog::new().pick_file() {
+                                self.encrypt.keyfile_path = path.as_path().display().to_string();
+                            }
+                        }
+                    });
+                }
+
+                if self.encrypt.key == Key::AutoGenerate {
+                    ui.horizontal(|ui| {
+                        ui.label("Passphrase: ");
+                        ui.add(egui::TextEdit::singleline(
+                            &mut self.encrypt.autogenerated_passphrase,
+                        ));
+                    });
+                }
+
+                if ui.button("Encrypt File").clicked() {
+                    // encrypty stuff, move to separate function
+                    let params = self.encrypt.clone();
+
+                    // discard the error as the user will be alerted to the issue
+                    let _ = std::thread::spawn(move || {
+                        let stor = std::sync::Arc::new(domain::storage::FileStorage);
+                        let input_file = stor
+                            .read_file(params.input_path.clone())
+                            .map_err(|_| message_box("Unable to read the input file."))
+                            .unwrap(); // !!!!
```

Comment:
You can return `Option<T>` from thread... and use the following:
```
stor.read_file(...).map_err(|_| message_box("...")).ok()?;
```

...or use the following

```
let input_file = stor.read_file(...) {
   Ok(input_file) = input_file,
   _ => return message_box("...")
}
```

Reply:
Good call, I couldn't make my mind up on how I wanted to do it.

The code is still heavily a WIP and will change rapidly, so I'm not _too_ concerned about it yet :)

Reply:
I know it's still in Draft, but it just confused me that you would show an error and the app would panic)

Reply:
By the way, we could make a macro that will do the second option) because I have doubts about the first one... "will it show an message box?"

Reply:
I have tested, and it does show the message box with the error, and then panics that thread.

We ideally need a way to show the message box, and then close the worker thread (instead of letting it panic).

As it's just the thread that panics, the app still works and it allows the user to continue (the worker thread panics silently in the background).